### PR TITLE
fix: add nil checks for controller and host keeper services

### PR DESCRIPTION
**Pull request by colin-axner**
## Description

closes: #XXXX

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer
- [ ] Review `Codecov Report` in the comment section below once CI passes

### CODE REVIEW

File: modules/apps/27-interchain-accounts/controller/keeper/migrations.go [hunk 1]
```diff
@@ -22,19 +22,21 @@ func NewMigrator(keeper *Keeper) Migrator {
 // AssertChannelCapabilityMigrations checks that all channel capabilities generated using the interchain accounts controller port prefix
 // are owned by the controller submodule and ibc.
 func (m Migrator) AssertChannelCapabilityMigrations(ctx sdk.Context) error {
-	for _, ch := range m.keeper.GetAllActiveChannels(ctx) {
-		name := host.ChannelCapabilityPath(ch.PortId, ch.ChannelId)
-		cap, found := m.keeper.scopedKeeper.GetCapability(ctx, name)
-		if !found {
-			return sdkerrors.Wrapf(capabilitytypes.ErrCapabilityNotFound, "failed to find capability: %s", name)
+	if m.keeper != nil {
```
Comment (colin-axner) [inline modules/apps/27-interchain-accounts/controller/keeper/migrations.go:25]: This ensures a no-op migration for chains which don't use the controller module

File: modules/apps/27-interchain-accounts/module.go [hunk 2]
```diff
@@ -150,9 +151,14 @@ func (am AppModule) LegacyQuerierHandler(legacyQuerierCdc *codec.LegacyAmino) sd
 
 // RegisterServices registers module services
 func (am AppModule) RegisterServices(cfg module.Configurator) {
-	controllertypes.RegisterMsgServer(cfg.MsgServer(), controllerkeeper.NewMsgServerImpl(am.controllerKeeper))
-	controllertypes.RegisterQueryServer(cfg.QueryServer(), am.controllerKeeper)
-	hosttypes.RegisterQueryServer(cfg.QueryServer(), am.hostKeeper)
+	if am.controllerKeeper != nil {
+		controllertypes.RegisterMsgServer(cfg.MsgServer(), controllerkeeper.NewMsgServerImpl(am.controllerKeeper))
+		controllertypes.RegisterQueryServer(cfg.QueryServer(), am.controllerKeeper)
```
Comment (colin-axner) [inline modules/apps/27-interchain-accounts/module.go:156]: It isn't possible to unregister the cli commands without modifying the AppModuleBasic struct

**Review by AdityaSripal — APPROVED** [review]
utACK

**Review by damiannolan — APPROVED** [review]

**Review by crodriguezvega — COMMENTED** [review]
Thanks, @colin-axner.

Is there any tests that we could add for this?

File: CHANGELOG.md [hunk 3]
```diff
@@ -72,6 +72,7 @@ Ref: https://keepachangelog.com/en/1.0.0/
 
 ### Improvements
 
+* (27-interchain-accounts) [\#2308](https://github.com/cosmos/ibc-go/pull/2308) Nil checks have been added to ensure services are not registered for nil host or controller keepers.
```
[the remark this answers is not in the document]
Comment (colin-axner) [reply CHANGELOG.md:75]: good point

**Comment by colin-axner** [comment]
> Is there any tests that we could add for this?

This is a bit tricky. I think ideally we would have e2e tests which do queries against controller and host submodules so chains which don't integrate both would run into a bug if we didn't have this fix. It is hard to test with go unit tests without modifying app.go

**Review by chatton — APPROVED** [review]
LGTM

**Comment by BlueTDeadly** [comment]
This was done expeditiously, can I assume that we saw this as a legitimate security concern? 

I just see this as work done in a silo without relation to the vulnerability or the WHY this was done so quickly. Any information would be helpful

**Comment by 0xekez** [comment]
@BlueTDeadly the commit that fixes the Quicksilver vulnerability seems to be this one:

https://github.com/cosmos/ibc-go/commit/4b3d3b8ae182bc56d56b9cc74e4ab794dd4a29e2

i personally would very much appreciate if patches like that would be linked in release notes. it may be useful for people reviewing the recent [cosmos hub proposal](https://www.mintscan.io/cosmos/proposals/97) to upgrade, for example.

**Comment by colin-axner** [comment]
Hi, we appreciate your feedback and we are very sorry to hear about exploitation of a patched issue. The original releases, [`v3.3`](https://github.com/cosmos/ibc-go/releases/tag/v3.3.0), and [`v4.1`](https://github.com/cosmos/ibc-go/releases/tag/v4.1.0) do indicate that a vulnerability was patched:

> This release includes a fix for a security vulnerability identified in the ICS27 host submodule. User funds in accounts and staking delegations are NOT at risk.

There was also a post on the [IBC security telegram channel](https://t.me/ibc_is_expansive) stating:
> We have released ibc-go v3.3.0 and v4.1.0 with a fix for a security vulnerability in the ICA host submodule. If your chain uses ICA host functionality, we recommend to upgrade as soon as possible. Feel free to contact me privately for more details or support.

Are there other locations we could link the security patch that would have increased visibility on the need to update the ica host module?

**Pull request merged**